allennlp/interpret/attackers/input_reduction.py

```diff
@@ -0,0 +1,144 @@
+class InputReduction(Attacker):
+    """
+    Runs the input reduction method from `Pathologies of Neural Models Make Interpretations
+    Difficult` https://arxiv.org/abs/1804.07781, which removes as many words as possible
+    from the input _without_ changing the model's prediction.
+    """
+    def attack_from_json(self, inputs: JsonDict,
+                         name_of_input_field_to_attack: str = 'tokens',
+                         name_of_grad_input_field: str = 'grad_input_1',
+                         ignore_tokens: List[str] = ["@@NULL@@"]): # pylint: disable=dangerous-default-value
+        original_instances = self.predictor.inputs_to_labeled_instances(inputs)
+        final_tokens = []
+        fields_to_check = {}
+        for current_instances in original_instances:
+            current_instances = [current_instances]
+            original_tokens = \
+                [x for x in current_instances[0][name_of_input_field_to_attack].tokens]
+
+            # Save fields that must be checked for equality
+            test_instances = self.predictor.inputs_to_labeled_instances(inputs)
+            for key in current_instances[0].fields.keys():
+                if key not in inputs and key != name_of_input_field_to_attack:
+                    fields_to_check[key] = test_instances[0][key]
+
+            # Set num_ignore_tokens, which tells input reduction when to stop
+            num_ignore_tokens = 0
+            # Keep at least one token for classification/entailment/etc.
+            if "tags" not in current_instances[0]:
+                num_ignore_tokens = 1
+
+            # Set num_ignore_tokens for NER and build token mask
+            else:
+                num_ignore_tokens, tag_mask, original_tags = \
+                    get_ner_tags_and_mask(current_instances, name_of_input_field_to_attack, ignore_tokens)
+            current_tokens = current_instances[0][name_of_input_field_to_attack].tokens
+            smallest_idx = -1
+            # keep removing tokens
+            while len(current_instances[0][name_of_input_field_to_attack]) >= num_ignore_tokens:
+                # get gradients and predictions
+                grads, outputs = self.predictor.get_gradients(current_instances)
+                for output in outputs:
+                    if isinstance(outputs[output], torch.Tensor):
+                        outputs[output] = outputs[output].detach().cpu().numpy().squeeze().squeeze()
+                    elif isinstance(outputs[output], list):
+                        outputs[output] = outputs[output][0]
+
+                # Check if any fields have changed, if so, break loop
+                if "tags" not in current_instances[0]:
+                    equal = True
+                    for field in fields_to_check:
+                        if field in current_instances[0].fields:
+                            equal = equal and current_instances[0][field].__eq__(fields_to_check[field])
```

Comment:
A bunch of similar comments here as in hotflip; is there a way to share any of this code between the two?